server/s3_object_metadata.cc

```diff
@@ -35,126 +37,123 @@
 
 extern struct m0_uint128 global_instance_id;
 
-void S3ObjectMetadata::initialize(bool ismultipart, std::string uploadid) {
-  json_parsing_error = false;
+S3ObjectMetadata::S3ObjectMetadata(const S3ObjectMetadata& src)
+    : S3ObjectMetadataCopyable(src), state(S3ObjectMetadataState::present) {
+
+  s3_log(S3_LOG_DEBUG, request_id, "Partial copy constructor");
+
+  bucket_name = request->get_bucket_name();
+  object_name = request->get_object_name();
+
+  initialize();
+}
+
+static void str_set_default(std::string& sref, const char* sz) {
+  if (sref.empty() && sz) {
+    sref = sz;
+  }
+}
+
+void S3ObjectMetadata::initialize() {
+
   account_name = request->get_account_name();
   account_id = request->get_account_id();
   user_name = request->get_user_name();
   canonical_id = request->get_canonical_id();
   user_id = request->get_user_id();
-  state = S3ObjectMetadataState::empty;
-  is_multipart = ismultipart;
-  upload_id = uploadid;
-  oid = M0_ID_APP;
-  old_oid = {0ULL, 0ULL};
-  layout_id = 0;
-  old_layout_id = 0;
-  tried_count = 0;
 
   object_key_uri = bucket_name + "\\" + object_name;
 
   // Set the defaults
-  system_defined_attribute["Date"] = "";
-  system_defined_attribute["Content-Length"] = "";
-  system_defined_attribute["Last-Modified"] = "";
-  system_defined_attribute["Content-MD5"] = "";
+  (void)system_defined_attribute["Date"];
+  (void)system_defined_attribute["Content-Length"];
+  (void)system_defined_attribute["Last-Modified"];
+  (void)system_defined_attribute["Content-MD5"];
+
   // Initially set to requestor of API, but on load json can be overriden
   system_defined_attribute["Owner-User"] = user_name;
   system_defined_attribute["Owner-Canonical-id"] = canonical_id;
   system_defined_attribute["Owner-User-id"] = user_id;
   system_defined_attribute["Owner-Account"] = account_name;
   system_defined_attribute["Owner-Account-id"] = account_id;
 
-  system_defined_attribute["x-amz-server-side-encryption"] =
-      "None";  // Valid values aws:kms, AES256
-  system_defined_attribute["x-amz-version-id"] =
-      "";  // Generate if versioning enabled
-  system_defined_attribute["x-amz-storage-class"] =
-      "STANDARD";  // Valid Values: STANDARD | STANDARD_IA | REDUCED_REDUNDANCY
-  system_defined_attribute["x-amz-website-redirect-location"] = "None";
-  system_defined_attribute["x-amz-server-side-encryption-aws-kms-key-id"] = "";
-  system_defined_attribute["x-amz-server-side-encryption-customer-algorithm"] =
-      "";
-  system_defined_attribute["x-amz-server-side-encryption-customer-key"] = "";
-  system_defined_attribute["x-amz-server-side-encryption-customer-key-MD5"] =
-      "";
+  str_set_default(system_defined_attribute["x-amz-server-side-encryption"],
+                  "None");
+  (void)system_defined_attribute["x-amz-version-id"];
+
+  // Generate if versioning enabled
+  str_set_default(system_defined_attribute["x-amz-storage-class"], "STANDARD");
```

Comment:
why these functions ?
instead of explicit initialization

Reply:
This function is called for both empty instance and copied instance.
We must avoid overwriting copied values.

Reply:
std::map<>::operator[] creates the element if it does not exist